### ISSUE TITLE
working with streams

### ISSUE THREAD

**Issue by danmayer** [karls/collage#8]
is there a way to work with an image stream and opposed to writing to file getting the file stream... Basically looking to pull a bunch of images on S3 convert and write back to S3... writing to disk both the read and the converted file to write back up to S3 seems a bit silly, but I am not sure if I could use this without having the temp files.

**Comment by karls** [comment]
Hi Dan,

Excellent question! It's been a while since I've done any serious Clojure. It's not possible to use collage to communicate directly with S3 at the moment. I suspect you might be able to use something like weavejester's clj-aws-s3 library to get and put S3 objects, and use collage to transform them in-memory.

For example, as a very rough sketch, you could

1. Use `get-object` from clj-aws-s3 to get the image from S3. You'll probably have to add an implementation for the `as-image` method in `ImageResource` protocol for `InputStream` (see https://github.com/karls/collage/blob/master/src/fivetonine/collage/util.clj#L82).
2. Do all the transformations on the image you need using collage.
3. Use `put-object` from clj-aws-s3 to write the in-memory image back to S3. This, again, will probably need a separate function to get the in-memory image data into a suitable shape for writing back to S3. It looks like you might be able to an implementation for `ToPutRequest` for `BufferedImage` (see https://github.com/weavejester/clj-aws-s3/blob/master/src/aws/sdk/s3.clj#L134).

Does this make sense? Does this help you at all?

**Comment by danmayer** [comment]
yeah sorry my question isn't clear, already using weavejester's clj-aws-s3 which is great, right now I have the image and can do the transform... but I have been pulling with his library, writing to file, then using this library to convert it and then writing that to file, then using clj-aws-s3 to write it back to S3. 

Cool so it sounds like if I added a method to as-image to work with input streams, I might be able to do transforms on streams and avoid the temporary files. Thanks that is really helpful. I really appreciate the response.

**Comment by karls** [comment]
Yeah, gotcha. I totally agree, it would be a lot better (and cooler!) if writing to files could be avoided. I think you should be able to using protocols both in collage (`ImageResource`) and clj-aws-s3 (`ToPutRequest`).

I'm a bit tight on time, but let me know if you get stuck, I might be able to help. Otherwise, I think it's a nice puzzle to piece together and learn some Clojure in the process. :)

**Comment by danmayer** [comment]
yeah should be good, I will take a look and see about send a PR your way.
